src/main/java/com/epam/aidial/core/service/ApplicationService.java

```diff
@@ -279,4 +489,420 @@ private void prepareApplication(ResourceDescription resource, EtagHeader etag,
+
+            if (request != null) {
+                request.reset();
+            }
+
+            if (e instanceof ExecutionException) {
+                e = e.getCause();
+            }
+
+            throw e;
+        }
+    }
+
+    private String deploymentLockKey(ResourceDescription resource) {
+        return BlobStorageUtil.toStoragePath(lockService.getPrefix(), "deployment:" + resource.getAbsoluteFilePath());
+    }
+
+    private String encodeTargetFolder(ResourceDescription resource, String id) {
+        String location = resource.getBucketLocation()
+                          + DEPLOYMENTS_NAME + PATH_SEPARATOR
+                          + id + PATH_SEPARATOR;
+
+        String name = encryptionService.encrypt(location);
+        return ResourceDescription.fromDecoded(ResourceType.FILE, name, location, null).getUrl();
+    }
+
+    public static boolean isActive(Application application) {
+        return application != null && application.getFunction() != null && isActive(application.getFunction().getStatus());
+    }
+
+    private static boolean isActive(Application.Function.Status status) {
+        return switch (status) {
+            case CREATED, FAILED, STOPPED -> false;
+            case STARTING, STARTED, STOPPING -> true;
+        };
+    }
+
+    private static boolean isPending(Application application) {
+        return application != null && application.getFunction() != null && isPending(application.getFunction().getStatus());
+    }
+
+    private static boolean isPending(Application.Function.Status status) {
+        return switch (status) {
+            case CREATED, STARTED, FAILED, STOPPED -> false;
+            case STARTING, STOPPING -> true;
+        };
+    }
+
+    private static void verifyApplication(ResourceDescription resource) {
+        if (resource.isFolder() || resource.getType() != ResourceType.APPLICATION) {
+            throw new IllegalArgumentException("Invalid application url: " + resource.getUrl());
+        }
+    }
+
+    private void verifyController() {
+        if (controllerEndpoint == null) {
+            throw new HttpException(HttpStatus.SERVICE_UNAVAILABLE, "The functionality is not available");
+        }
+    }
+
+    private static void verifyMapping(String path, boolean required, String message) {
+        if (path == null) {
+            if (required) {
+                throw new IllegalArgumentException(message);
+            }
+
+            return;
+        }
+
+        if (!path.startsWith("/")) {
+            throw new IllegalArgumentException(message);
+        }
+
+        try {
+            UrlUtil.decodePath(path, true);
+        } catch (Throwable e) {
+            throw new IllegalArgumentException(message);
+        }
+    }
+
+    private static String buildMapping(String endpoint, String path) {
+        return (endpoint == null || path == null) ? null : (endpoint + path);
+    }
+
+    private static <T> T convertServerSentEvent(String body, Class<T> clazz) {
```

Comment:
Let's move to utility class and cover with tests

Reply:
Already covered. Moved.